### PR TITLE
[python] add type hints in docs/conf.py

### PR DESCRIPTION
**Pull request by jameslamb**
Adds type hints to code in `docs/conf.py`, as part of #3756.

### Notes for Reviewers

#3756 has been open for about 8 months now, so I think it's ok for maintainers to start picking up the work to finish it and not wait any longer for outside contributors to find it from searches for `good first issue`.

Better type hints will improve the ability of `mypy` to catch issues (#3868 ).

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -197,7 +198,7 @@ def run(self):
 latex_logo = str(CURR_PATH / 'logo' / 'LightGBM_logo_black_text_small.png')
 
 
-def generate_doxygen_xml(app):
+def generate_doxygen_xml(app: Any) -> None:
```
Comment (StrikerRUS) [inline docs/conf.py:201]: Maybe use `sphinx.application.Sphinx` for `app` (here and at L246 and L308)?
https://www.sphinx-doc.org/en/master/extdev/appapi.html

Comment (jameslamb) [reply docs/conf.py:201]: Sure, can do. I was just following the documentation below.

```text
    Parameters
    ----------
    app : object
        The application object representing the Sphinx process.
```

Comment (StrikerRUS) [reply docs/conf.py:201]: I think we can update docstring as well.

Comment (jameslamb) [reply docs/conf.py]: sure! Updated in https://github.com/microsoft/LightGBM/pull/4526/commits/81f0a8259078df7abe12f92b0bcee811129db127

File: docs/conf.py [hunk 2]
```diff
@@ -197,12 +199,12 @@ def run(self):
...
 
 
-def generate_doxygen_xml(app):
+def generate_doxygen_xml(app: Sphinx) -> None:
     """Generate XML documentation for C API by Doxygen.
 
     Parameters
     ----------
-    app : object
+    app : Sphinx
```
Comment (StrikerRUS) [inline docs/conf.py:207]: I think we can be more precise here.
```suggestion
    app : sphinx.application.Sphinx
```

File: docs/conf.py [hunk 3]
```diff
@@ -242,12 +244,12 @@ def generate_doxygen_xml(app):
...
 
 
-def generate_r_docs(app):
+def generate_r_docs(app: Sphinx) -> None:
     """Generate documentation for R-package.
 
     Parameters
     ----------
-    app : object
+    app : Sphinx
```
Comment (StrikerRUS) [inline docs/conf.py:252]:
```suggestion
    app : sphinx.application.Sphinx
```

File: docs/conf.py [hunk 4]
```diff
@@ -304,12 +306,12 @@ def generate_r_docs(app):
...
 
 
-def setup(app):
+def setup(app: Sphinx) -> None:
     """Add new elements at Sphinx initialization time.
 
     Parameters
     ----------
-    app : object
+    app : Sphinx
```
Comment (StrikerRUS) [inline docs/conf.py:314]:
```suggestion
    app : sphinx.application.Sphinx
```

**Review by StrikerRUS — APPROVED** [review]
LGTM!

**Pull request merged**